File: scripts/xval_arb.py
#!/usr/bin/env python3
"""Sortie machine du detecteur Python, a confronter au C++."""
from __future__ import annotations
import sys
from pathlib import Path
sys.path.insert(0, str(Path(__file__).resolve().parents[1]))

from kairos.arb import CostModel, Detector, DeviationBasis, PairQuote

PAIRES = [
    ("EUR","USD",1.08000,1.08004), ("USD","JPY",157.000,157.006),
    ("GBP","USD",1.27150,1.27160), ("EUR","GBP",0.84930,0.84937),
    ("USD","CHF",0.88120,0.88127), ("AUD","USD",0.66410,0.66417),
    ("EUR","JPY",169.562,169.572), ("GBP","JPY",199.640,199.652),
]
d = Detector()
for b,q,bid,ask in PAIRES:
    d.add_pair(PairQuote(b,q,bid,ask))

print(f"n_devises {len(d.currencies)}")
print(f"n_paires {len(PAIRES)}")
for k in (3,4,5):
    print(f"cycles_{k} {len(d.enumerate_cycles(k))}")

c = CostModel(0.20, 2.00, 0.0)
for base in (DeviationBasis.EXECUTABLE, DeviationBasis.MID):
    for notional in (10_000, 100_000):
        opps = d.scan(c, notional, 3, 5, False, base)
        tag = f"{base.value}_{notional}"
        print(f"{tag}_n {len(opps)}")
        print(f"{tag}_best_gross {opps[0].gross_bps:.17e}")
        print(f"{tag}_best_cost {opps[0].cost_bps:.17e}")
        print(f"{tag}_best_net {opps[0].net_bps:.17e}")
        print(f"{tag}_sum_net {sum(o.net_bps for o in opps):.17e}")
        print(f"{tag}_worst_net {opps[-1].net_bps:.17e}")

l = d.resolve("GBP","USD"); print(f"gbpusd_rate {l.rate:.17e}")
l = d.resolve("USD","GBP"); print(f"usdgbp_rate {l.rate:.17e}")

File: kairos-cpp/apps/xval_arb.cpp
// Pendant C++ de scripts/xval_arb.py. Memes entrees, memes grandeurs.
#include "kairos/arb/detector.hpp"
#include <cstdio>
using namespace kairos::arb;

int main() {
    const struct { const char *b, *q; double bid, ask; } P[] = {
        {"EUR","USD",1.08000,1.08004}, {"USD","JPY",157.000,157.006},
        {"GBP","USD",1.27150,1.27160}, {"EUR","GBP",0.84930,0.84937},
        {"USD","CHF",0.88120,0.88127}, {"AUD","USD",0.66410,0.66417},
        {"EUR","JPY",169.562,169.572}, {"GBP","JPY",199.640,199.652}};
    Detector d;
    for (const auto& p : P) d.add_pair({p.b, p.q, p.bid, p.ask});

    std::printf("n_devises %zu\n", d.n_currencies());
    std::printf("n_paires %zu\n", d.n_pairs());
    for (int k = 3; k <= 5; ++k)
        std::printf("cycles_%d %zu\n", k, d.enumerate_cycles(k).size());

    const CostModel c{0.20, 2.00, 0.0};
    const struct { DeviationBasis b; const char* n; } bases[] = {
        {DeviationBasis::Executable, "executable"}, {DeviationBasis::Mid, "mid"}};
    for (const auto& bs : bases)
        for (double notional : {10'000.0, 100'000.0}) {
            const auto o = d.scan(c, notional, 3, 5, false, bs.b);
            char tag[64];
            std::snprintf(tag, sizeof tag, "%s_%.0f", bs.n, notional);
            std::printf("%s_n %zu\n", tag, o.size());
            std::printf("%s_best_gross %.17e\n", tag, o.front().gross_bps);
            std::printf("%s_best_cost %.17e\n", tag, o.front().cost_bps);
            std::printf("%s_best_net %.17e\n", tag, o.front().net_bps);
            double s = 0; for (const auto& x : o) s += x.net_bps;
            std::printf("%s_sum_net %.17e\n", tag, s);
            std::printf("%s_worst_net %.17e\n", tag, o.back().net_bps);
        }
    std::printf("gbpusd_rate %.17e\n", d.resolve("GBP","USD")->rate);
    std::printf("usdgbp_rate %.17e\n", d.resolve("USD","GBP")->rate);
    return 0;
}
